### ISSUE TITLE
Refactor rename

### ISSUE THREAD

**Issue by vaukalak** [facebook/nuclide#688]
### Issue and Steps to Reproduce

There's no possibility to execute refactor-rename operation.

### Expected Behavior
If the project is covered by flow, nuclide should find references to a class, property, file, method and rename them safely.

### Actual Behavior
Seems like feature is not implemented.

### Versions
* Atom: 1.10.0
* Nuclide: 0.162.0

**Comment by nmote** [comment]
We're working on this, but currently Flow does not support find references or rename operations. The Flow team needs to build that, and we need to build the UI.

**Comment by gisenberg** [comment]
@nmote Is it possible to get a pointer to an issue on Flow that tracks the missing implementation?

**Comment by nmote** [comment]
Don't think there is one, and the work isn't planned. We'd like to do it at some point but we don't yet know when.

**Comment by vaukalak** [comment]
[https://github.com/facebook/flow/releases/tag/v0.38.0](https://github.com/facebook/flow/releases/tag/v0.38.0 )

```
Local find references command: flow find-refs file line column - 
if there is a variable x at file:line:column,
then this command returns the location of x's 
definition and the location of all reads and writes to x.
```

**Comment by gisenberg** [comment]
@nmote It looks like this issue can be reopened based on the flow release notes for v0.38.0.  Thoughts?

**Comment by vaukalak** [comment]
@gisenberg yes and this one probably https://github.com/facebook/nuclide/issues/689

**Comment by nmote** [comment]
Nope, check out my answer here: https://github.com/facebook/nuclide/issues/689#issuecomment-281215820

The current Flow find-refs is a good first step but it's woefully inadequate for any real refactoring functionality. And even the limited functionality that it promises is currently really buggy.

**Comment by FDiskas** [comment]
Ubuntu, by pressing Ctrl+Alt+R does nothing on typescript file :(

![refactor](https://i.imgur.com/8qcPD3w.png)

Nuclide v: 0.280.0
Atom v: 1.23.3 x64
:snail:

**Comment by ip** [comment]
So are there plans to support this very common refactoring? VS Code has this out of the box with TypeScript.

![image](https://user-images.githubusercontent.com/1111008/36283569-33e3403e-12a5-11e8-82b8-9961328dd178.png)

**Comment by scottwio** [comment]
How are facebook handling this internally? I would imagine on large codebases refactoring of this kind is fairly common.

**Comment by nmote** [comment]
I've been working on global `find-refs` in Flow. It's currently in beta and works fairly well, though there are still a number of known issues. You can try it out in Nuclide by checking the "Enable global find references" setting under the Flow category. Make sure you are using at least Flow v0.66.0 because there are some pretty serious issues with `find-refs` before that version.

Over the next couple months I plan to implement global rename atop `find-references`. This will be exposed through the refactorizer command in Nuclide that @FDiskas screenshotted above.

@scottwio I'm sure that you can imagine how this is handled currently. Obviously it's less than ideal so I am looking forward to shipping rename.

**Comment by scottwio** [comment]
Any movement on this it's been a couple of months starting to really miss this feature from typescript

**Comment by nmote** [comment]
Yeah, I've been continuing to make `find-refs` more reliable, which will directly impact the quality of rename, and I plan in the next 4-6 weeks to get a prototype for rename landed. I'll probably gate it behind a setting in Nuclide at first while I work out the kinks but if you're interested you'll be able to access it. I would use care at first to make sure it's doing the right thing.
